lib/shoulda/matchers/active_record/have_attached_matcher.rb

```diff
@@ -0,0 +1,133 @@
+module Shoulda
+  module Matchers
+    module ActiveRecord
+      def have_one_attached(name)
+        HaveAttachedMatcher.new(:one, name)
+      end
+
+      def have_many_attached(name)
+        HaveAttachedMatcher.new(:many, name)
+      end
+
+      # @private
+      class HaveAttachedMatcher
+        attr_reader :name
+
+        def initialize(macro, name)
+          @macro = macro
+          @name = name
+        end
+
+         def description
+          "has_#{macro}_attached #{name}"
+        end
+
+        def failure_message
+          "Expected #{expectation} (#{@failure})"
+        end
+
+        def failure_message_when_negated
+          "Did not expect #{expectation}"
+        end
+
+        def expectation
+          "#{model_class.name} to have a has_#{macro} attached called #{name}"
+        end
+
+        def matches?(subject)
+          @subject = subject
+          reader_attribute_exists? &&
+            writer_attribute_exists? &&
+            attachments_association_exists? &&
+            blobs_association_exists? &&
+            eager_loading_scope_exists?
+        end
+
+        private
+
+        attr_reader :subject, :macro
+
+        def attachments_association_exists?
+          if attachments_association_matcher.matches?(subject)
+            true
+          else
+            @failure = "no valid association called #{attachments_association_name}"
+            false
+          end
+        end
+
+        def attachments_association_matcher
+          @attachments_association_matcher ||=
+            AssociationMatcher.new(:"has_#{macro}", attachments_association_name).
+              conditions(name: name).
+              class_name('ActiveStorage::Attachment').
+              inverse_of(:record).
+              dependent(false)
+        end
+
+        def blobs_association_exists?
+          if blobs_association_matcher.matches?(subject)
+            true
+          else
+            @failure = "no valid association called #{blobs_association_name}"
```

Comment:
As mentioned above, what do you think about reusing the message that the matcher generates here? Either `blobs_association_matcher.failure_message` or `blobs_association_matcher.failure_message_when_negated`, depending on how _this_ matcher is called.